File: sources/07-hash-table/03-hash-table-with-indirect-addressing-and-flexible-linked-list/c/Main.c
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>

/**
 * Código fonte - Implementação da Tabela Hash Indireta com Lista Simples Flexível
 *
 * @author Artur Bomtempo Colen
 * @algorithm Data Structure
 * @type Hash Table with Indirect Addressing and Flexible Linked List
 * @language C
 * @version 1.0
 */

typedef struct Cell {
    int element;
    struct Cell *next;
} Cell;

typedef struct List {
    Cell *first;
    Cell *last;
} List;

void initList(List *list) {
    list->first = NULL;
    list->last = NULL;
}

void insertStart(List *list, int value) {
    Cell *temp = (Cell *)malloc(sizeof(Cell));
    temp->element = value;
    temp->next = list->first;
    list->first = temp;

    if (list->last == NULL) {
        list->last = temp;
    }
}

void insertEnd(List *list, int value) {
    Cell *temp = (Cell *)malloc(sizeof(Cell));
    temp->element = value;
    temp->next = NULL;

    if (list->last == NULL) {
        list->first = temp;
        list->last = temp;
    } else {
        list->last->next = temp;
        list->last = temp;
    }
}

int removeStart(List *list) {
    if (list->first == NULL) {
        printf("Erro ao remover: Lista vazia\n");
        return -1;
    }

    Cell *temp = list->first;
    int element = temp->element;
    list->first = list->first->next;

    if (list->first == NULL) {
        list->last = NULL;
    }

    free(temp);

    return element;
}

int removeEnd(List *list) {
    if (list->first == NULL) {
        printf("Erro ao remover: Lista vazia\n");
        return -1;
    }

    Cell *temp = list->first;

    if (temp->next == NULL) {
        int element = temp->element;
        free(temp);
        list->first = NULL;
        list->last = NULL;
        return element;
    }

    while (temp->next != list->last) {
        temp = temp->next;
    }

    int element = list->last->element;
    free(list->last);
    list->last = temp;
    list->last->next = NULL;

    return element;
}

bool removeByValue(List *list, int value) {
    if (list->first == NULL) {
        printf("Erro ao remover: Lista vazia\n");
        return false;
    }

    Cell *prev = NULL;
    Cell *current = list->first;

    while (current != NULL) {
        if (current->element == value) {
            if (prev == NULL) {
                list->first = current->next;
                if (list->first == NULL) {
                    list->last = NULL;
                }
            } else {
                prev->next = current->next;
                if (current == list->last) {
                    list->last = prev;
                }
            }

            free(current);

            return true;
        }

        prev = current;
        current = current->next;
    }

    return false;
}

bool search(List *list, int value) {
    Cell *current = list->first;

    while (current != NULL) {
        if (current->element == value) {
            return true;
        }

        current = current->next;
    }

    return false;
}

void show(List *list) {
    Cell *current = list->first;
    printf("[");

    while (current != NULL) {
        printf("%d ", current->element);
        current = current->next;
    }

    printf("]\n");
}

typedef struct Hash {
    List *table;
    int size;
} Hash;

void initHash(Hash *hash, int size) {
    hash->size = size;
    hash->table = (List *)malloc(sizeof(List) * size);

    for (int i = 0; i < size; i++) {
        initList(&hash->table[i]);
    }
}

int hashFunction(Hash *hash, int value) { return value % hash->size; }

void insertStartHash(Hash *hash, int value) {
    int index = hashFunction(hash, value);
    insertStart(&hash->table[index], value);
}

bool searchHash(Hash *hash, int value) {
    int index = hashFunction(hash, value);
    return search(&hash->table[index], value);
}

bool removeHash(Hash *hash, int value) {
    int index = hashFunction(hash, value);
    return removeByValue(&hash->table[index], value);
}

int main() {
    Hash hashTable;
    initHash(&hashTable, 7);

    printf("Inserindo valores...\n");
    insertStartHash(&hashTable, 10);
    insertStartHash(&hashTable, 23);
    insertStartHash(&hashTable, 36);
    insertStartHash(&hashTable, 49);
    insertStartHash(&hashTable, 15);
    insertStartHash(&hashTable, 64);

    printf("\nTabela Hash após inserções:\n");
    for (int i = 0; i < 7; i++) {
        printf("Índice %d: ", i);
        show(&hashTable.table[i]);
    }

    printf("\nPesquisando valores...\n");
    printf("Valor 23 encontrado? %s\n",
           searchHash(&hashTable, 23) ? "Sim" : "Não");
    printf("Valor 99 encontrado? %s\n",
           searchHash(&hashTable, 99) ? "Sim" : "Não");

    printf("\nRemovendo valores...\n");
    printf("Valor 23 removido? %s\n",
           removeHash(&hashTable, 23) ? "Sim" : "Não");
    printf("Valor 99 removido? %s\n",
           removeHash(&hashTable, 99) ? "Sim" : "Não");

    printf("\nTabela Hash após remoções:\n");
    for (int i = 0; i < 7; i++) {
        printf("Índice %d: ", i);
        show(&hashTable.table[i]);
    }

    return 0;
}

File: sources/07-hash-table/03-hash-table-with-indirect-addressing-and-flexible-linked-list/java/Main.java
/**
 * Código fonte - Implementação da Tabela Hash Indireta com Lista Simples Flexível
 *
 * @author Artur Bomtempo Colen
 * @algorithm Data Structure
 * @type Hash Table with Indirect Addressing and Flexible Linked List
 * @language Java
 * @version 1.0
 */

class Cell {
    public int element;
    public Cell next;

    public Cell() {
        this.element = 0;
        this.next = null;
    }

    public Cell(int value) {
        this.element = value;
        this.next = null;
    }
}

class List {
    private Cell first, last;

    public List() {
        this.first = new Cell();
        last = first;
    }

    public void insertStart(int value) {
        Cell temp = new Cell(value);

        temp.next = first.next;
        first.next = temp;

        if (first == last) {
            last = temp;
        }

        temp = null;
    }

    public void insertEnd(int value) {
        last.next = new Cell(value);
        last = last.next;
    }

    public void insert(int value, int position) {
        int length = length();

        if (position < 0 || position > length) {
            throw new Error("Não foi possível inserir valor na lista.");
        } else if (position == 0) {
            insertStart(value);
        } else if (position == length) {
            insertEnd(value);
        } else {
            Cell i = first;

            for (int j = 0; j < position; j++, i = i.next)
                ;

            Cell temp = new Cell(value);

            temp.next = i.next;
            i.next = temp;

            temp = i = null;
        }
    }

    public int removeStart() {
        if (first == last) {
            throw new Error("Não foi possível remover um valor da lista.");
        }

        Cell temp = first.next;
        first.next = first.next.next;

        int element = temp.element;

        temp.next = null;
        temp = null;

        return element;
    }

    public int removeEnd() {
        if (first == last) {
            throw new Error("Não foi possível remover um valor da lista.");
        }

        Cell i;

        for (i = first; i.next != last; i = i.next)
            ;

        int element = i.element;

        last = i;
        i = last.next = null;

        return element;
    }

    public int remove(int value, int position) {
        int element, length = length();

        if (position < 0 || position > length) {
            throw new Error("Não foi possível remover um valor da lista.");
        } else if (position == 0) {
            element = removeStart();
        } else if (position == length) {
            element = removeEnd();
        } else {
            Cell i = first;

            for (int j = 0; j < position; j++, i = i.next)
                ;

            Cell temp = i.next;

            element = temp.element;
            i.next = temp.next;

            temp.next = null;
            i = temp = null;
        }

        return element;
    }

    public boolean removeByValue(int value) {
        boolean removed = false;

        Cell prev = first;

        for (Cell i = first.next; i != null; prev = i, i = i.next) {
            if (i.element == value) {
                prev.next = i.next;

                if (i == last) {
                    last = prev;
                }

                i.next = null;
                removed = true;
                break;
            }
        }

        return removed;
    }

    public int length() {
        int length = 0;
        for (Cell i = first; i != last; i = i.next, length++)
            ;
        return length;
    }

    public boolean search(int value) {
        boolean found = false;

        for (Cell i = first.next; i != null; i = i.next) {
            if (i.element == value) {
                found = true;
                i = last;
            }
        }

        return found;
    }

    public void show() {
        System.out.print("[");

        for (Cell i = first.next; i != null; i = i.next) {
            System.out.print(i.element + " ");
        }

        System.out.print("]");
    }
}

class Hash {
    List table[];
    int size;

    public Hash() {
        this.size = 7;
    }

    public Hash(int size) {
        this.size = size;
        this.table = new List[size];

        for (int i = 0; i < size; i++) {
            table[i] = new List();
        }
    }

    public int h(int value) {
        return value % size;
    }

    public void inserStart(int value) {
        int i = h(value);
        table[i].insertStart(value);
    }

    public boolean search(int value) {
        int i = h(value);
        return table[i].search(value);
    }

    public boolean removal(int value) {
        int i = h(value);
        boolean removed = table[i].removeByValue(value);

        return removed;
    }
}

public class Main {
    public static void main(String[] args) {
        Hash hashTable = new Hash(7);

        System.out.println("Inserindo valores...");
        hashTable.inserStart(10);
        hashTable.inserStart(23);
        hashTable.inserStart(36);
        hashTable.inserStart(49);
        hashTable.inserStart(15);
        hashTable.inserStart(64);

        System.out.println("\nTabela Hash após inserções:");
        for (int i = 0; i < 7; i++) {
            System.out.print("Índice " + i + ": ");
            hashTable.table[i].show();
            System.out.println();
        }

        System.out.println("\nPesquisando valores...");
        System.out.println("Valor 23 encontrado? " + hashTable.search(23));
        System.out.println("Valor 99 encontrado? " + hashTable.search(99));

        System.out.println("\nRemovendo valores...");
        System.out.println("Valor 23 removido? " + hashTable.removal(23));
        System.out.println("Valor 99 removido? " + hashTable.removal(99));

        System.out.println("\nTabela Hash após remoções:");
        for (int i = 0; i < 7; i++) {
            System.out.print("Índice " + i + ": ");
            hashTable.table[i].show();
            System.out.println();
        }
    }
}